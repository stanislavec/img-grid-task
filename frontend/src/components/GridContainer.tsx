import React, { Fragment } from 'react';
import scss from './GridContainer.module.scss';

interface IContainer<T> {
  isCheckedToBeEmpty?: boolean;
  items: T[];
  renderItem(item: T): JSX.Element;
  getKey(item: T): string | number;
}

function GridContainer<T = unknown>(props: IContainer<T>) {
  const { items, renderItem, getKey, isCheckedToBeEmpty = true } = props;

  if (!renderItem || typeof renderItem !== 'function')
    throw Error('Render function is not passed as a prop');

  if (!getKey || typeof getKey !== 'function')
    throw Error('GetKey function is not passed as a prop');

  if (!items.length && isCheckedToBeEmpty) return <div className={scss.empty}>No data...</div>;

  return (
    <div className={scss.container}>
      {items.map((item) => (
        <Fragment key={getKey(item)}>{renderItem(item)}</Fragment>
      ))}
    </div>
  );
}

export default GridContainer;
