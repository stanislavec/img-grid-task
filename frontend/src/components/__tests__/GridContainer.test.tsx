import { render } from '@testing-library/react';
import GridContainer from '../GridContainer';

const items = [1, 2, 3, 4];

describe('GridContainer', () => {
  it('is defined', () => {
    const elem = render(
      <GridContainer items={items} renderItem={(item) => <p>{item}</p>} getKey={(item) => item} />,
    );
    expect(elem.baseElement).toBeDefined();
  });

  it('is throws error without required props', () => {
    try {
      // @ts-ignore
      render(<GridContainer items={items} />);
    } catch (error: any) {
      expect(error).toBeInstanceOf(Error);
      expect(error.message).toBe('Render function is not passed as a prop');
    }

    try {
      // @ts-ignore
      render(<GridContainer items={items} renderItem={() => null} />);
    } catch (error: any) {
      expect(error).toBeInstanceOf(Error);
      expect(error.message).toBe('GetKey function is not passed as a prop');
    }
  });

  it('renders correctly with empty items object', () => {
    const elem = render(
      <GridContainer items={[]} renderItem={(item) => <p>{item}</p>} getKey={(item) => item} />,
    );
    expect(elem.baseElement).toHaveTextContent('No data...');
  });

  it('is matches snapshot with empty result', () => {
    const elem = render(
      <GridContainer items={[]} renderItem={(item) => <p>{item}</p>} getKey={(item) => item} />,
    );
    expect(elem.container).toMatchSnapshot();
  });

  it('is matches snapshot', () => {
    const elem = render(
      <GridContainer items={items} renderItem={(item) => <p>{item}</p>} getKey={(item) => item} />,
    );
    expect(elem.container).toMatchSnapshot();
  });
});
