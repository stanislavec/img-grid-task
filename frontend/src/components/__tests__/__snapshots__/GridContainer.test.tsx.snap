// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridContainer is matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <p>
      1
    </p>
    <p>
      2
    </p>
    <p>
      3
    </p>
    <p>
      4
    </p>
  </div>
</div>
`;

exports[`GridContainer is matches snapshot with empty result 1`] = `
<div>
  <div
    class="empty"
  >
    No data...
  </div>
</div>
`;
